Add SWDP driver interface API and bit-bang driver, add CMSIS-DAP compatible controller

jfischer-no opened the pull request:
Add Serial Wire Debug Port interface driver API and bit-bang driver.
Add CMSIS-DAP compatible controller which is a handler between
the host interface and SWD driver.
Add cmsis dap sample using ~~USB HID as interface~~ bulk transfers.

This is a rebase and followup of #20183

jfischer-no committed:
driver: add SWDP driver interface API and bit-bang driver

Add Serial Wire Debug Port interface driver API and bit-bang driver.

The driver requires a simple Hardware Interface Circuits (HICs),
where signals CLK, DOUT, DIN, ENn, OE_ENn, RESETn
are connected to board GPIOs and buffered signals SWD_CLK and SWD_DIO
to the target.

Signal OE_ENn controls the direction of the Serial Wire (SWD_DIO),
ENn the buffers SWD_CLK possibly others and enables/disables HIC.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dap: add CMSIS-DAP compatible controller

Add CMSIS-DAP compatible controller which is a handler between
the host interface and SWD driver. The controller follows CMSIS-DAP
reference implementation. It expects a request buffer from the host
interface, splits it to simple transfers and forwards to the DP driver,
and finally returns a response buffer to the host.
Interface to the host can be implemented with USB HID device support.

Controller implements only SW-DP support and is tested
with pyOCD and ADIv5.x.

Signed-off-by: Johann Fischer <j.fischer@phytec.de>

jfischer-no committed:
samples: debug: add CMSIS DAP sample using USB as interface

Add CMSIS DAP sample using USB as interface.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

pdgendt commented on drivers/dp/swdp_bitbang.c:
```diff
@@ -0,0 +1,632 @@
+	LOG_HEXDUMP_DBG(sw_request_lut, sizeof(sw_request_lut), "request lut");
+}
+
+static uint32_t ALWAYS_INLINE sw_get32bit_parity(uint32_t data)
```
Nit: Keep the order consistent with the other functions.
```suggestion
static ALWAYS_INLINE uint32_t sw_get32bit_parity(uint32_t data)
```

pdgendt commented on drivers/dp/swdp_bitbang.c:
```diff
@@ -0,0 +1,632 @@
+		*response = (uint8_t)ack;
+	}
+
+	return 0;
```
Should  `0` be the result in case of a protocol error?

jfischer-no replied:
Yes, protocol error is handled using ack/response, so we can use return value for controller IO errors.

pdgendt commented on drivers/dp/swdp_bitbang.c:
```diff
@@ -0,0 +1,632 @@
+#endif
+}
+
+static ALWAYS_INLINE void pin_platform_set(void *base, uint8_t pin)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+	NRF_GPIO_Type * reg = base;
+
+	reg->OUTSET = BIT(pin);
+#else
+#error "Not defined for this SoC family"
+#endif
+}
```
I'm a bit confused by these platform specific functions. Does every SoC we want to support need to be added here? Can't we abstract this using existing drivers + devicetree?

nordicjm replied:
Would assume it is for speed, you need these GPIO transitions to be fast, using the zephyr drivers would be universal but probably add far too much overhead/delay. There is a description of this in the DAPLINK repository.

jfischer-no replied:
Yes, it is what @nordicjm says. I made it more generic as in my previous PR and throughput has worsened. But bit-bang driver is rather for initial work and verification, we have at least two platforms where much faster implementation using specific peripherals is possible.

rerickson1 replied:
Would be cool to have a sample on the RP2040 using the PIO system.

MaureenHelm commented:
@flit fyi

fabiobaltieri commented on drivers/dp/swdp_bitbang.c:
```diff
@@ -0,0 +1,632 @@
+static ALWAYS_INLINE void pin_delay_asm(uint32_t delay)
+{
+#if defined(CONFIG_SOC_SERIES_NRF52X)
+	__asm volatile ("movs r3, %[p]\n"
+			".start_%=:\n"
+			"subs r3, #1\n"
+			"bne .start_%=\n"
+			:
+			: [p] "r" (delay)
+			: "r3", "cc"
+			);
```
Wouldn't this work with any Cortex-M? Maybe it could be a `#ifdef CONFIG_CPU_CORTEX_M`

cfriedt replied:
@jfischer-no, @fabiobaltieri - I did something similar to this for `fpga_ice40.c`. it can likely be done without resorting to assembly, which makes it slightly more portable.

In the case for the ice40, the DT bindings include a parameter for calibrated delay for exactly a 1 MHz square wave. Maybe I should use the pattern below as well, to allow the frequency to be configurable on a per-system basis.

My suggestions / requests for this:
* an additional DT binding for a calibration table indexed by frequency
  * key: frequency in Hz, val: delay count
  * sorted in increasing order, which would permit both linear and binary search in the driver
  * maybe make that an inline "helper" function in a private header, similar to gpio helpers
* driver API to query the calibration table keys (and therefore min, max, other supported frequencies)
* driver API to query the current frequency
* driver API to set the SWD frequency by frequency instead of count

The reason I'm leaning towards frequency in the API is because it is maybe a bit clearer for users.

Something to consider is whether the `set_frequency` API call would only support exact frequencies or clamping to the nearest frequency. That sort of thing should be consistent between any driver implementation though, so would need to be verified in a testsuite.

Slightly related: https://github.com/zephyrproject-rtos/zephyr/pull/54947#discussion_r1138567407

https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/fpga/lattice%2Cice40-fpga.yaml
https://github.com/zephyrproject-rtos/zephyr/blob/bb6fa5707f77d72307008405f19dd83247405392/drivers/fpga/fpga_ice40.c#L132

tmon-nordic replied:
While I agree that it is nice to have the timing parameter provided by device tree, I don't like following:
- The delay is to be calibrated for CPU running the code, not ice40. Note sure if such property should be part of `lattice,ice40-fpga` compatible or rather more related to the core running the code.
- Calibrating for loop that does only handle clock but using it for loop that handles both clock and data. The achieved frequency in actual loop is going to be slower.

Doing a instruction processing based delays is going to be really problematic the more complex CPU this runs on, i.e. the most repeatable results you get on a system without a cache, but the more sophisticated instruction cache or pipeline there is the more annoying it is going to get.

raveslave replied:
did anyone experiment with chained dma descriptors for speed-up?
or even adding an extra input-pin -> IRQ on the swdio, the target ack could be masked on tx, but used as a trigger for ack to speed things up?

nice if it's cpu agnostic of course even though I like the inline asm idea above

seems many run LPC's for their debug probes, these have a Programmable Logic Unit (PLU) feature that could do some simple counting logic

jfischer-no replied:
> Do we already have an umbrella `zephyr,swdp.yaml`? If not, eventually it might make sense to create one so that `zephyr,swdp-gpio.yaml` and e.g. a PIO implementation could have a common base

That can be done later, I am looking forward #44316

fabiobaltieri commented on drivers/dp/swdp_bitbang.c:
```diff
@@ -0,0 +1,632 @@
+
+	reg->OUTSET = BIT(pin);
+#else
+#error "Not defined for this SoC family"
```
Could you fallback to using the GPIO APIs automatically?

nordicjm replied:
GPIO API is too slow

fabiobaltieri replied:
Is it so slow that it does not work at all? I mean if it's just slow it may be better than not working on at all, you could add a big #warning about it of course.

Also @cfriedt I remember seeing something about using native GPIO for bitbanging in one of your FPGA PRs as well, maybe we could have some optional common macros for that as part of the GPIO APIs?

tmon-nordic replied:
[ARM®DS-5 Version 5 ARM DSTREAM System and Interface Design Reference Guide Table 2-2 SWD timing requirements](https://developer.arm.com/documentation/100956/0529/Arm-DSTREAM-Target-Interface-Connections/Signal-descriptions/Serial-Wire-Debug?lang=en) specifies Tlow and Thigh min as 10 ns and max as 500 us. Note that the 500 us is the absolute maximum you have between SWCLK pulses including all overhead.

Note that the 500 us is the max you have from ARM itself, the minimum frequency required can be higher depending on target device. For example for [nRF51 Series](https://infocenter.nordicsemi.com/pdf/nRF51_RM_v3.0.pdf) the minimum SWDCLK is 125 kHz which translates to 8 us max (instead of 500 us).

fabiobaltieri replied:
Wow, was about to answer something like "we can't do 2MHz with the standard API?!" but then I hacked an infinite `gpio_pin_toggle_dt()` on an nrf52dk and got a mere 860kHz. Well, thanks for explaining.

jfischer-no replied:
Fallback to GPIO is now implemented.

nordicjm reviewed (commented):
Was not able to get this working with an nrf52840dk connected to an nrf52832, ACK timed out for every request. Re-wired the nrf52832 directly to a DAPLINK programmer (without nRESET) and it works fine through that using pyocd.

cfriedt commented on dts/bindings/misc/zephyr,swdp-gpio.yaml:
```diff
@@ -0,0 +1,45 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: >
+    This is a representation of the SW GPIO interface
```
Might make sense to expand SW to "Single Wire Debug (SWD)"

jfischer-no commented:
Rebased. I have not addressed all the comments yet, will do so later.

raveslave commented:
shouldn't this make use of CMSIS-DAP v2 (USB bulk) instead of the HID?  
or at least provide an option to choose as larger packages and high-speed usb is important for speed

> Use [CMSIS-DAP v2.x](https://arm-software.github.io/CMSIS_5/DAP/html/group__DAP__ConfigUSB__gr.html) instead that provides high-speed SWO trace streaming and does not require driver installation in modern operating systems (Mac OS, Linux, Windows).

jfischer-no commented:
> shouldn't this make use of CMSIS-DAP v2 (USB bulk) instead of HID?
> or at least provide an option to choose bulk/v2 vs hid as larger packages and high-speed usb is important for speed

No. USB transfer is only an option, it is not the main scope of this PR, it is there for demonstration and testing. Bit-bang controller performance is far far away from the possible USB FS interrupt transfers maximum bandwidth. I see no need to discuss this until we have stable API with few fast SWDP controllers. Also, no one forces you to use sample in this PR, you are free to implement own host backend.

raveslave commented:
> > shouldn't this make use of CMSIS-DAP v2 (USB bulk) instead of HID?
> > or at least provide an option to choose bulk/v2 vs hid as larger packages and high-speed usb is important for speed
> 
> No. USB transfer is only an option, it is not the main scope of this PR, it is there for demonstration and testing. Bit-bang controller performance is far far away from the possible USB FS interrupt transfers maximum bandwidth. 

yes, true for swd communication which needs a mcu-specific optimization for speed.

but for vendor-commands running over daplink you can make use of the packet-size and added speed of **v2** compared to HID, even when using FS phy's

flit commented:
Regardless of performance, I strongly recommend switching the example/test to use CMSIS-DAP v2 / bulk endpoints. HID for CMSIS-DAP is deprecated and should no longer be used for new code or examples.

Fwiw, bit bang SWD is able to achieve 2-10 MHz, sometimes even higher, depending on the MCU and how well the code is implemented (eg, direct register writes with constant register addresses and bit masks). However, the duty cycle tends to be all over the place (not a problem protocol-wise, but ugly as hell on a scope 😝). Point being that on some controller MCUs, you can achieve faster SWD performance with bulk.

Commit:
dap: add Konfig option to set maximum packet size

This also fixes a bug where the packet size returned was only 17 byte.

Signed-off-by: Johan Carlsson <johan.carlsson@teenage.engineering>

Commit:
dap: do not return error in DAP connect command

If a host client crashes or fails to disconnect, the probe will
hang. Allow a new connection without raising errors.

Signed-off-by: Johan Carlsson <johan.carlsson@teenage.engineering>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

te-johan commented:
Great work @jfischer-no! I've tested to integrate it into a probe we have and it was rather simple to get it going. Added support for nxp lpc series and could run it with 512 byte packet size on high speed usb.
I could not get HID to work but i repurposed another usb class driver(midi) I had to get the CMSIS-DAPv2 bulk interface to work.

please feel free to integrate them into your rebase branch. I did not include the usb class driver since i do not have it in the zephyr tree.
https://github.com/te-johan/zephyr/tree/cmsis-dap-lpc-fixes

maxd-nordic commented:
FYI: I'm currently working on updating this to CMSIS-DAPv2. Generic device support and simpler PIN configurations are already there. I based it on @te-johan 's version.
https://github.com/maxd-nordic/sdk-zephyr/tree/pr-cmsis-dap-rebase-additions

te-johan commented:
I can upstream my cmsis-dapv2 class driver next week.
I’ve also fixed some reset issues as well as bumped the specification to 1.3 to support raspberry pico.
It can easily be lifted to 2.1.1 since all new features are optional.

maxd-nordic commented:
@te-johan I'm sure we can find a way to merge our additions.

maxd-nordic commented:
This is a PR with my additions: https://github.com/jfischer-no/zephyr/pull/2
Unfortunately, there is some bug with block-transfers that I couldn't fix yet.
Any help is highly appreciated! @jfischer-no @tmon-nordic @te-johan

maxd-nordic committed:
driver: swdp_bitbang: rework pin configurations

Move low-level GPIO functions to a separate file and use GPIO driver
API if low-level GPIO support is not available for the platform.
Allows alternative pin configuration using only two pins, clk and dio.
Improve binding description.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

maxd-nordic committed:
dap: implement wait for SWJ pins command

Implement wait for SWJ pins command.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

maxd-nordic committed:
dap: add support for DAP_INFO string elements

Add support for DAP_INFO string elements.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

jfischer-no committed:
drivers: swdp_bitbang: add SW output sequence

Add API to read count bits from SWDIO into data LSB first.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dap: react properly to unsupported UART commands

Add ID_DAP_UART_* command definitions and react properly to
unsupported UART commands.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dap: implement DAP SWD sequence command

DAP SWD sequence command is a requirement to support CMSIS-DAPv2.
Raise supported version to "2.1.0".

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented:
Rebased and applied patches (with little and few with heavy rework) from @te-johan and @maxd-nordic.
The driver now supports different pin configurations and will fall back to GPIO API if the platform is not supported.
Sample USB interface  to the host is implemented using bulk endpoints.

maxd-nordic commented:
@jfischer-no The string descriptor is missing now. pyocd will not detect this without it.

henrikbrixandersen commented on subsys/dap/cmsis_dap.c:
```diff
@@ -0,0 +1,1062 @@
+		api->swdp_set_pins(ctx->swdp_dev, select, value);
+	}
+
+	/* FIXME */
```
Does this comment still apply?

henrikbrixandersen commented on drivers/dp/swdp_ll_pin.h:
```diff
@@ -0,0 +1,82 @@
+#define CPU_CLOCK		CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC
+#endif
+
+#if defined(CONFIG_SOC_SERIES_NRF52X) || defined(CONFIG_SOC_SERIES_NRF53X)
+#define FAST_BITBANG_HW_SUPPORT 1
+#else
+#define FAST_BITBANG_HW_SUPPORT 0
+#endif
```
Having one header file to encapsulate all future low-level bitbang implementations likely wont scale?

jfischer-no replied:
It is okay for now (82 lines), we will see how it evolves and react. But implementing faster GPIO for all platforms is not our goal here.

rerickson1 commented on dts/bindings/misc/zephyr,swdp-gpio.yaml:
```diff
@@ -0,0 +1,46 @@
+    noe-gpios:
+      type: phandle-array
+      required: true
+      description: GPIO pin used for SWD NOE output
```
Can't these be optional. There are many hardware designs that do not need to use input/output buffers

henrikbrixandersen reviewed (commented):
A little more documentation of the architecture (the split between the DAP subsystem, the debug port driver, the bitbang and the optimized GPIO access) wouldn't hurt.

Apart from that, I think the architecture of this looks sane for an initial, experimental version. I have added a few questions below:

rerickson1 commented on subsys/dap/cmsis_dap.c:
```diff
@@ -0,0 +1,1062 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * This file is based on DAP.c from DAPLink Interface Firmware.
+ * Copyright (c) 2009-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
It might be nice to reference the commit SHA that this was started from. This way we can more easily compare if updates are made to the original ARM file.
Also, put a link to the original repo.

jfischer-no replied:
Checked it against my presentation and updated accordingly, added version and link, no SHA.

rerickson1 commented on subsys/dap/cmsis_dap.c:
```diff
@@ -0,0 +1,1062 @@
+	return retval;
+}
+
+/* Process DAP Vendor command request */
+static uint16_t dap_process_vendor_cmd(struct dap_context *const ctx,
+				       const uint8_t *const request,
+				       uint8_t *const response)
+{
+	response[0] = ID_DAP_INVALID;
+	return 1U;
+}
```
Define this as weak so a user can easily override this in their app. That way a user doesn't have to modify code in the zephyr tree.

jfischer-no replied:
No, there are other ways to achieve it.

rerickson1 replied:
of course, open to better ways

jfischer-no replied:
This has not been addressed yet, I know how, but I do not have the time yet.

rerickson1 replied:
Would like to see this done before this PR is merged.

cfriedt replied:
It should be relatively easy to register even a single callback, and that could be done in a follow-up PR. Vendor commands are relatively specialized as well, so why add it in a generic PR?

There are probably other reasons to not block here, maybe creeping into the neighborhood of policy.. 🤷🏼‍♂️

rerickson1 replied:
Nice point, adding it later is fine. I was trying to push my own agenda because this is a key feature I need :)

henrikbrixandersen commented:
@jfischer-no Is the target milestone still v3.6.0?

aescolar commented:
Adding DNM after talkign with @jfischer-no as he prefered it not merged yet, so as to avoid it being merged as it meets the release criteria. @jfischer-no please remove the label whenever you want.

jfischer-no commented on include/zephyr/drivers/swdp.h:
This is common driver API header, like `include/zephyr/drivers/uart.h `

jfischer-no commented on include/zephyr/drivers/swdp.h:
```diff
@@ -0,0 +1,97 @@
+ * This is the mandatory API any Serial Wire driver needs to expose.
+ */
+struct swdp_api {
+	/* Generate SWJ Sequence according to sequence bit count and bit data */
```
This has not been addressed yet, I did not have time for it yesterday, pushed it anyway, maybe I can fix it later tonight.

te-johan commented:
I added vendor commands a while back. feel free to include.
https://github.com/zephyrproject-rtos/zephyr/commit/7f7e50f4b028844b23e5c274c59aa55b5c565cb0

mniestroj commented on subsys/dap/cmsis_dap.c:
```diff
@@ -0,0 +1,1062 @@
+	uint8_t value = request[0];
+	uint8_t select = request[1];
+	uint32_t wait = sys_get_le32(&request[2]);
+	int64_t timeout_ticks;
+	uint8_t state;
+
+	if (!atomic_test_bit(&ctx->state, DAP_STATE_CONNECTED)) {
+		LOG_ERR("DAP device is not connected");
+		response[0] = DAP_ERROR;
+		return 1U;
+	}
+
+	/* Skip if nothing selected. */
+	if (select) {
+		api->swdp_set_pins(ctx->swdp_dev, select, value);
+	}
+
+	timeout_ticks = k_uptime_ticks() + (CONFIG_SYS_CLOCK_TICKS_PER_SEC / 1000000 * wait);
+
+	do {
+		api->swdp_get_pins(ctx->swdp_dev, &state);
+		LOG_INF("select 0x%02x, value 0x%02x, wait %u, state 0x%02x",
+			select, value, wait, state);
+		if ((value & select) == (state & select)) {
+			LOG_DBG("swdp_get_pins succeeded before timeout");
+			break;
+		}
+	} while (k_uptime_ticks() - timeout_ticks > 0);
```
```suggestion
	} while (k_uptime_ticks() < timeout_ticks);
```
Otherwise it never leaves the loop.

maxd-nordic replied:
maybe like this?
```suggestion
	k_timepoint_t end = sys_timepoint_calc(K_USEC(wait));
	uint8_t state;

	if (!atomic_test_bit(&ctx->state, DAP_STATE_CONNECTED)) {
		LOG_ERR("DAP device is not connected");
		response[0] = DAP_ERROR;
		return 1U;
	}

	/* Skip if nothing selected. */
	if (select) {
		api->swdp_set_pins(ctx->swdp_dev, select, value);
	}

	do {
		api->swdp_get_pins(ctx->swdp_dev, &state);
		LOG_INF("select 0x%02x, value 0x%02x, wait %u, state 0x%02x",
			select, value, wait, state);
		if ((value & select) == (state & select)) {
			LOG_DBG("swdp_get_pins succeeded before timeout");
			break;
		}
	} while (!sys_timepoint_expired(end));
```

mniestroj commented on subsys/dap/cmsis_dap.c:
```diff
@@ -0,0 +1,1062 @@
+		api->swdp_set_pins(ctx->swdp_dev, select, value);
+	}
+
+	timeout_ticks = k_uptime_ticks() + (CONFIG_SYS_CLOCK_TICKS_PER_SEC / 1000000 * wait);
```
```suggestion
	timeout_ticks = k_uptime_ticks() + (CONFIG_SYS_CLOCK_TICKS_PER_SEC * wait / 1000000);
```
When `CONFIG_SYS_CLOCK_TICKS_PER_SEC=32768` dividing it by `1000000` will always result in 0. We should first multiply, then divide.

pdgendt replied:
Maybe this can be calculated using the `k_timepoint_t` instead?

mniestroj replied:
Great idea! I was not aware of this API :)

mniestroj commented on subsys/dap/cmsis_dap.h:
```diff
@@ -0,0 +1,136 @@
+#define DAP_TRANSFER_MISMATCH			BIT(4)
+
+/* DAP controller capabilities */
+#define DAP_DP_SUPPOTS_SWD			BIT(0)
+#define DAP_DP_SUPPOTS_JTAG			BIT(1)
+#define DAP_SWO_SUPPOTS_UART			BIT(2)
+#define DAP_SWO_SUPPOTS_MANCHESTER		BIT(3)
+#define DAP_SUPPORTS_ATOMIC_COMMANDS		BIT(4)
+#define DAP_SUPPORTS_TIMESTAMP_CLOCK		BIT(5)
+#define DAP_SWO_SUPPOTS_STREAM			BIT(6)
```
```suggestion
#define DAP_DP_SUPPORTS_SWD			BIT(0)
#define DAP_DP_SUPPORTS_JTAG			BIT(1)
#define DAP_SWO_SUPPORTS_UART			BIT(2)
#define DAP_SWO_SUPPORTS_MANCHESTER		BIT(3)
#define DAP_SUPPORTS_ATOMIC_COMMANDS		BIT(4)
#define DAP_SUPPORTS_TIMESTAMP_CLOCK		BIT(5)
#define DAP_SWO_SUPPORTS_STREAM			BIT(6)
```

maxd-nordic committed:
dap: fix timeout handling

Use k_timepoint_t for timeout handling in the swj_pins function.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

maxd-nordic committed:
drivers: swdp: document API

This patch adds documentation for the SWDP API.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>
Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

maxd-nordic committed:
driver: swdp_bitbang: hardcode request lut

Hardcode the lookup table for SWDP requests.
This is an optimization to save some space.
Documentation was added to understand the values.

Signed-off-by: Maximilian Deubel <maximilian.deubel@nordicsemi.no>

maxd-nordic commented on include/zephyr/drivers/swdp.h:
```diff
@@ -0,0 +1,151 @@
+	 * @param turnaround Line turnaround cycles
+	 * @param data_phase Always generate Data Phase (also on WAIT/FAULT)
+	 * @return 0 on success, or error code
+	*/
```
for compliance:
```suggestion
	 */
```

maxd-nordic commented on samples/subsys/dap/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+    build_only: true
+    platform_allow: nrf52840dk_nrf52840
+    integration_platforms:
+      - nrf52840dk_nrf52840
```
```suggestion
      - nrf52840dk/nrf52840
```

maxd-nordic replied:
(occurs multiple times)

jfischer-no replied:
I reworked sample configuration and removed thingy91 which is only supported in NCS.

maxd-nordic commented on drivers/dp/swdp_bitbang.c:
```diff
@@ -0,0 +1,632 @@
+	}
+
+	LOG_HEXDUMP_DBG(sw_request_lut, sizeof(sw_request_lut), "request lut");
+}
```
this was done with the latest changes 👍🏻
